### PR TITLE
Support for converting underscore literals

### LINKED ISSUE #46: Support underscores in mpz for string input
Should work exactly as for int.  (Note that gmpy2 allows more permissible syntax.)

### PR DESCRIPTION
**Pull request by rruuaanng**
Closes #46

### CODE REVIEW

**Comment by rruuaanng** [comment]
cc @skirpichev

**Review by skirpichev — CHANGES_REQUESTED** [review]
I don't understand why you mark this as ready for review.

Obviously, your test just check arithmetic ops for random integers.  We have such tests in other functions.

Your code is invalid, it accepts more input, e.g.:
```pycon
>>> mpz('1__23')
mpz(123)
>>> int('1__23')
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    int('1__23')
    ~~~^^^^^^^^^
ValueError: invalid literal for int() with base 10: '1__23'
```
Please read issue first.  You should replicate int()'s behavior.  Please read CPython docs: it's described what's a correct string input for int() and what's not.

**Comment by rruuaanng** [comment]
I don't seem to have tested the high input. I will improve it.

**Comment by rruuaanng** [comment]
It's ok. Look this
```
>>> mpz('123__1_')
Traceback (most recent call last):
  File "<python-input-8>", line 1, in <module>
    mpz('123__1_')
    ~~~^^^^^^^^^^^
ValueError: invalid literal for mpz() with base 10: '123__1_'
>>> int('123__1_')
Traceback (most recent call last):
  File "<python-input-9>", line 1, in <module>
    int('123__1_')
    ~~~^^^^^^^^^^^
ValueError: invalid literal for int() with base 10: '123__1_'
```

**Comment by rruuaanng** [comment]
More test
```
ValueError: invalid literal for mpz() with base 10: '_-1'
Traceback (most recent call last):
  File "<python-input-24>", line 1, in <module>
    mpz('_-1');int('_-1')
               ~~~^^^^^^^
ValueError: invalid literal for int() with base 10: '_-1'
>>> mpz('-1_');int('-1_')
Traceback (most recent call last):
  File "<python-input-25>", line 1, in <module>
    mpz('-1_');int('-1_')
    ~~~^^^^^^^
ValueError: invalid literal for mpz() with base 10: '-1_'
Traceback (most recent call last):
  File "<python-input-25>", line 1, in <module>
    mpz('-1_');int('-1_')
               ~~~^^^^^^^
ValueError: invalid literal for int() with base 10: '-1_'
>>> mpz('-1_1');int('-1_1')
-11
>>> mpz('-1_1_');int('-1_1_')
Traceback (most recent call last):
  File "<python-input-27>", line 1, in <module>
    mpz('-1_1_');int('-1_1_')
    ~~~^^^^^^^^^
ValueError: invalid literal for mpz() with base 10: '-1_1_'
Traceback (most recent call last):
  File "<python-input-27>", line 1, in <module>
    mpz('-1_1_');int('-1_1_')
                 ~~~^^^^^^^^^
ValueError: invalid literal for int() with base 10: '-1_1_'
>>> mpz('-1_1__');int('-1_1__')
Traceback (most recent call last):
  File "<python-input-28>", line 1, in <module>
    mpz('-1_1__');int('-1_1__')
    ~~~^^^^^^^^^^
ValueError: invalid literal for mpz() with base 10: '-1_1__'
Traceback (most recent call last):
  File "<python-input-28>", line 1, in <module>
    mpz('-1_1__');int('-1_1__')
                  ~~~^^^^^^^^^^
ValueError: invalid literal for int() with base 10: '-1_1__'
>>> mpz('1_1__');int('1_1__')
Traceback (most recent call last):
  File "<python-input-29>", line 1, in <module>
    mpz('1_1__');int('1_1__')
    ~~~^^^^^^^^^
ValueError: invalid literal for mpz() with base 10: '1_1__'
Traceback (most recent call last):
  File "<python-input-29>", line 1, in <module>
    mpz('1_1__');int('1_1__')
                 ~~~^^^^^^^^^
ValueError: invalid literal for int() with base 10: '1_1__'
>>> mpz('1_1__1');int('1_1__1')
Traceback (most recent call last):
  File "<python-input-30>", line 1, in <module>
    mpz('1_1__1');int('1_1__1')
    ~~~^^^^^^^^^^
ValueError: invalid literal for mpz() with base 10: '1_1__1'
Traceback (most recent call last):
  File "<python-input-30>", line 1, in <module>
    mpz('1_1__1');int('1_1__1')
                  ~~~^^^^^^^^^^
ValueError: invalid literal for int() with base 10: '1_1__1'
>>> mpz('1_1_1');int('1_1_1')
111
```

**Comment by rruuaanng** [comment]
I think it ready receive other review.

File: main.c [hunk 1]
```diff
@@ -239,6 +239,35 @@ const unsigned char gmp_digit_value_tab[] =
...
 
+int
+underscores_as_number(char *str, Py_ssize_t *len)
+{
+    int i, j, sep;
+    char *buf = PyMem_Malloc(sizeof(char) * (*len) + 1);
+
+    i = j = sep = 0;
+    if (*len == 0 || str == NULL || *str == '_') {
+        return 1;
```
Comment (skirpichev) [inline main.c:250]: return code now useless

File: main.c [hunk 2]
```diff
@@ -239,6 +239,35 @@ const unsigned char gmp_digit_value_tab[] =
...
+    }
+
```
Comment (skirpichev) [inline main.c:252]:
```suggestion
```

File: main.c [hunk 3]
```diff
@@ -239,6 +239,35 @@ const unsigned char gmp_digit_value_tab[] =
...
+            i++;
+            continue;
+        }
+        buf[j++] = str[i++];
+    }
+    *len -= sep;
+    memcpy(str, buf, *len + 1);
+    str[*len] = '\0';
+    PyMem_Free(buf);
+
```
Comment (skirpichev) [inline main.c:268]:
```suggestion
```

File: main.c [hunk 4]
```diff
@@ -251,6 +280,7 @@ MPZ_from_str(PyObject *s, int base)
 
     Py_ssize_t len;
     const char *str = PyUnicode_AsUTF8AndSize(s, &len);
+    underscores_as_number((char *)str, &len);
```
Comment (skirpichev) [inline main.c:283]: Put newline before and remove newline after.

Note also, that you introduce memory leak on failure.

File: main.c (hunk 1 above)
Comment (rruuaanng) [reply main.c:250]: Oh right. I'm modifying it.

File: main.c [hunk 5]
```diff
@@ -239,6 +239,48 @@ const unsigned char gmp_digit_value_tab[] =
   -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
 };
 
+int
```
Comment (skirpichev) [inline main.c:242]:
```suggestion

static int
```

File: main.c [hunk 6]
```diff
@@ -239,6 +239,48 @@ const unsigned char gmp_digit_value_tab[] =
...
+underscores_as_number(char *str, Py_ssize_t *len)
+{
+    int i, j, skip;
+    char next, prev;
+    char *buf = PyMem_Malloc(sizeof(char) * (*len) + 1);
```
Comment (skirpichev) [inline main.c:247]:
```suggestion
    char *buf = PyMem_Malloc(*len + 1);
```

File: main.c [hunk 7]
```diff
@@ -239,6 +239,48 @@ const unsigned char gmp_digit_value_tab[] =
...
+
+    i = j = skip = 0;
+    next = prev = '\0';
+    if (*len == 0 || str == NULL) {
+        goto error;
+    }
```
Comment (skirpichev) [inline main.c:253]: Please put declarations of variables, where they actually got used.

File: main.c [hunk 8]
```diff
@@ -239,6 +239,48 @@ const unsigned char gmp_digit_value_tab[] =
...
+            prev = str[i++];
+            skip++;
+            continue;
+        }
+        if (str[i] == '_') {
+            if ((next == '_' || next == ' ' || next == '\0')
+                || (prev == '_' || prev == ' ' || prev == '\0')) {
+                goto error;
+            }
+            prev = str[i++];
```
Comment (skirpichev) [inline main.c:267]:
```suggestion
            prev = str[i++];
```

File: main.c [hunk 9]
```diff
@@ -252,6 +294,7 @@ MPZ_from_str(PyObject *s, int base)
     Py_ssize_t len;
     const char *str = PyUnicode_AsUTF8AndSize(s, &len);
 
+    underscores_as_number((char *)str, &len);
```
Comment (skirpichev) [inline main.c:297]: You should check return value.

Also, please read Python docs about PyUnicode_* functions.  You did this in a wrong way, for example:
```pycon
>>> from gmp import mpz
>>> s = '1_2_3'
>>> mpz(s)
mpz(123)
>>> s
'123\x003'
```

Comment (rruuaanng) [reply main.c:297]: Hmmm, it seems that I missed something. I'll check them.

Comment (skirpichev) [reply main.c:297]: MPZ_from_str() code below already allocates a temporary buffer to handle prefixes, etc.  I would rather land such changes there.

Please also take look on the CPython code.

Comment (rruuaanng) [reply main.c:297]: I will carefully study the unicode implementation in CPython.

**Comment by rruuaanng** [comment]
I've been busy with the review of the Zephyr Devicetree subsystem recently. I've delayed it for too long. If you want, you can continue to push this PR.
(I'll finish it in my free time, but it's pretty exciting now that the function is almost done.) :-------)

**Comment by skirpichev** [comment]
Feel free to reopen this, when you feel it will be ready to review again.  Or start a new pr.

**Pull request closed without merging**